<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>SAAS: Search All, Assess Subset • saas</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">saas</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/index.html">Articles</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/compomics/search-all-assess-subset">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>SAAS: Search All, Assess Subset</h1>
                        <h4 class="author">Adriaan Sticker</h4>
            
            <h4 class="date">2017-04-21</h4>
          </div>

    
    
<div class="contents">
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>In this vignette we will demonstrate how to use the ‘search all, asses subset’ strategy (all-sub) to correctly calculated the FDR on a PSM subset of interest in a shotgun proteomic experiment.</p>
<p>In short, the basic all-sub workflow comprises the following steps:</p>
<ol style="list-style-type: decimal">
<li>Search the experimental MS2 spectra against all proteins potentially present in the sample.</li>
<li>Remove all PSMs that match a protein you are not interesed in.</li>
<li>Calculate the FDR on the resulting subset of PSMs.</li>
</ol>
<p>A preprint of the manuscript on the search-all-assess FDR procedure can be found on bioRxiv (<a href="https://doi.org/10.1101/094581" class="uri">https://doi.org/10.1101/094581</a>).</p>
<blockquote>
<p>Mass spectrometrists should search for all peptides, but assess only the ones they care about<br> Adriaan Sticker, Lieven Clement, Lennart Martens <span class="citation">(2017)</span></p>
</blockquote>
<div id="false-discovery-rate-fdr-estimation" class="section level2">
<h2 class="hasAnchor">
<a href="#false-discovery-rate-fdr-estimation" class="anchor"></a>False discovery rate (FDR) estimation</h2>
Let <span class="math inline">\(x\)</span> be the PSM score and assume that larger score values indicate a better match to the theoretical spectrum. The FDR is known to have a Bayesian interpretation upon defining a mixture model for the distribution of <span class="math inline">\(x\)</span> <span class="citation">(Efron 2008)</span>.
<span class="math display">\[\begin{equation}
f(x) = \pi_0 f_0(x) + (1-\pi_0) f_1(x),
\end{equation}\]</span>
<p>with <span class="math inline">\(f(x)\)</span> the target distribution of the test statistic, <span class="math inline">\(f_0(x)\)</span> the mixture component corresponding to incorrect PSMs, <span class="math inline">\(f_1(x)\)</span> the mixture component corresponding to the correct PSMs and <span class="math inline">\(\pi_0\)</span> the fraction of incorrect PSMs.</p>
Based on the mixture distribution the FDR can be defined as the posterior probability
<span class="math display">\[\begin{equation}
\text{FDR}(x) = Pr[\text{PSM is incorrect}\vert X\geq x].
\end{equation}\]</span>
Let <span class="math inline">\(F(x)\)</span>, <span class="math inline">\(F_0(x)\)</span> and <span class="math inline">\(F_1(x)\)</span> indicate the complementary cumulative distribution function (CCDF)
<span class="math display">\[\begin{equation}
F_{.}(x)=\int\limits_{X=x}^{+\infty} f_{.}(X) dx
\end{equation}\]</span>
so
<span class="math display">\[\begin{equation}
F(x)= \pi_0 F_0(x) + (1-\pi_0) F_1(x)
\end{equation}\]</span>
and the FDR becomes
<span class="math display">\[\begin{equation}
\text{FDR}(x)=\frac{\pi_0 F_0(x) }{F(x)}
\end{equation}\]</span>
<p>The empirical Bayes/two-groups approach also has the great virtue that independence of the scores is not necessary as long as (1) the estimated CCDF of the marginal distribution is roughly unbiased and (2) the null component is estimated empirically <span class="citation">(Efron 2008)</span>.</p>
</div>
<div id="target-decoy-approach" class="section level2">
<h2 class="hasAnchor">
<a href="#target-decoy-approach" class="anchor"></a>Target-decoy approach</h2>
<p>In a competitive target decoy search, FDRs are estimated by dividing the number of accepted decoys PSMs by the number of accepted target PSMs at a certain MS-GF+ score cutoff <span class="citation">(Elias and Gygi 2007)</span>. For FDR estimation in the subset of interest, use only target and decoy PSMs belonging to the subset in the calculation. We assure a monotonic increase of FDR estimates with decreasing score values by replacing the FDR at a particular score <span class="math inline">\(x\)</span> by the minimum TDA FDR in the interval <span class="math inline">\([-\infty,x]\)</span>.</p>
Note, that in a competitive target decoy search the FDR with TDA is estimated by
<span class="math display">\[\begin{equation}
\widehat{\text{FDR}}(x) =\frac{\# decoys | X\geq x}{\# targets | X\geq x},
\end{equation}\]</span>
which can be rewritten as 
<span class="math display">\[\begin{eqnarray*}
\widehat{\text{FDR}}(x) &amp;=&amp;\frac{\frac{\#decoys}{\#targets}\frac{\# decoys | X\geq x}{\# decoys}}{\frac{\# targets | X\geq x}{\# targets}}\\
&amp;=&amp;\frac{\hat \pi_0 \bar F_0(x)}{\bar F(x)}
\end{eqnarray*}\]</span>
with <span class="math inline">\(\bar F_0(x)\)</span> and <span class="math inline">\(\bar F(x)\)</span> the empirical complementary cumulative distribution functions (ECCDFs) estimated using decoy and target PSMs, respectively. So, in a competitive TDA the FDR implicitly relies on the estimator of the proportion of incorrect PSMs among the targets
<span class="math display">\[\begin{equation} \hat \pi_0=\frac{\#decoys}{\#targets}.
\end{equation}\]</span>
</div>
<div id="stable-target-decoy-approach-for-subsets" class="section level2">
<h2 class="hasAnchor">
<a href="#stable-target-decoy-approach-for-subsets" class="anchor"></a>Stable target-decoy approach for subsets</h2>
<p>In small subsets we typically can expect a low number of decoy PSMs and the <span class="math inline">\(\bar F_0(x)\)</span> will be a poor estimator of <span class="math inline">\(F_0(x)\)</span>. For improving the stability of the target-decoy approach in small subsets we propose to estimate the mixture component for incorrect PSMs using a large set of decoys with similar properties as the incorrect subset PSMs. Hence, the method only relies on the assumption that the incorrect target PSM scores follow the same distribution as the PSM scores of the decoy set.</p>
<p>This large set of decoys PSM scores can be designed by the user. However often, the set of decoys in the complete search seems a good candidate set. This allows for the estimation of the mixture component of the incorrect PSMs and the marginal distribution of all PSMs non-parametrically, i.e. based on the ECCDFs <span class="math inline">\(\bar F_0(x)\)</span> using a large decoy set (e.g. of the complete search) and <span class="math inline">\(\bar F(x)\)</span> derived from the targets in the subset, respectively.</p>
The mixture proportion <span class="math inline">\(\pi_0\)</span> can still be estimated based on the ratio of the number of decoys on the number of targets in the subset upon a competitive target decoy search. However, the estimator <span class="math inline">\(\hat \pi_0\)</span> often becomes zero in very small subsets, inducing an FDR estimate of zero for the set of all subset target PSMs. We therefore propose to use the estimator
<span class="math display">\[\begin{equation}
    \hat \pi_0=
\begin{cases}
  \frac{\#\text{decoys}_\text{subset}+1}{\#\text{targets}_\text{subset}}, &amp;
  \text{if}\ \#\text{targets} &gt; \#\text{decoys}\\
  1, &amp;
  \text{if}\ \#\text{targets} \leq \#\text{decoys}\\
\end{cases}
  ,
\end{equation}\]</span>
<p>which has a lower bound of <span class="math inline">\(\frac{1}{\text{targets}_\text{subset}}\)</span> and an upper bound of 1. Note that this estimator has an upward bias in small subsets, which disappears for moderate to large subsets. In the extreme case that we observe only one target PSM, <span class="math inline">\(\hat \pi_0\)</span> is 1 and the FDR estimator reduces to a regular p-value.</p>
<p>Note, that a conservative estimator of the FDR is also available using the link between the Bayesian and the Benjamini-Hochberg (BH) FDR procedure, i.e. by setting the mixture proportion of the mixture component for incorrect PSMs to <span class="math inline">\(\pi_0=1\)</span> <span class="citation">(Efron 2008)</span>.</p>
<span class="math display">\[\begin{equation}
\widehat{\text{FDR}}(x)=\frac{\bar F_0(x)}{\bar F(x)}.
\end{equation}\]</span>
<p>In very small subsets, the latter FDR estimator is still valid as long as a good estimator of the mixture component for incorrect PSMs is available. Benjamini and Hochberg, 1995, for instance, included an example with 15 comparisons in their seminal paper <span class="citation">(Benjamini and Hochberg 1995)</span>. Also note that their FDR estimator also reduces to a regular p-value in the extreme case that the subset of interest consists of one target PSM, i.e. we set <span class="math inline">\(\pi_0=1\)</span> and <span class="math inline">\(\bar F=1\)</span> for this extreme edge case.</p>
<p>The ECCDF <span class="math inline">\(\bar F_0(x)\)</span> will be zero for every target PSM with a score higher then the maximum decoy PSM score, inducing an FDR estimate of zero for these target PSMs. Decreasing the total number of decoys will lower the maximum decoy PSM score and thus lower the threshold at which the estimated FDR of the target PSMs will be zero. This means when only few decoys PSMs are observed, FDR estimates have high sample-to-sample variability and are biased to lower values.  FDR estimates are too liberal and underestimates the true FDR. Therefore we recommend to have at least 1000 decoys to obtain sufficiently reliable FDR estimates in most situations. Lastly, we assure a monotonic increase of FDR estimates with decreasing score values by replacing the FDR at a particular score <span class="math inline">\(x\)</span> by the minimum TDA FDR in the interval <span class="math inline">\([-\infty,x]\)</span>.</p>
</div>
</div>
<div id="example-data-analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#example-data-analysis" class="anchor"></a>Example data analysis</h1>
<div id="the-data" class="section level2">
<h2 class="hasAnchor">
<a href="#the-data" class="anchor"></a>The data</h2>
<p>In this vignette we use data from a <em>Pyrocococus furiosis</em> sample run on a LTQ-Orbitrap Velos mass spectrometer. The data can be found in the PRIDE repository with indentifier PXD001077. The <em>Pyrocococus furiosis</em> reference proteome fasta file was downloaded from Uniprot on April 22, 2016. In this use case we are interested in <em>Pyrococcus</em> proteins related to transferase activity as defined by their Gene Ontology identifier (<a href="GO:0016740" class="uri">GO:0016740</a>). A fasta file with these proteins was also downloaded from Uniprot.</p>
</div>
<div id="database-search" class="section level2">
<h2 class="hasAnchor">
<a href="#database-search" class="anchor"></a>Database search</h2>
<p>In theory, the all-sub method can be used on the results from any search engine that report peptide spectrum matches (PSMs) with a score. However, this R package only provides a parser for the MS-GF+ MZident output files. Output from other search engines should be parsed and presented in the format of the output from the ‘parse_msgf_mzid()’ function (see ‘help(saas::parse_msgf_mzid)’in R). It should be realtively straightforward to adapt the code from ’parse_msgf_mzid()’ to work with other search engines. It’s very important that searches are run on a concatenated target-decoy database.</p>
<p>The <em>Pyrococcus</em> dataset was searched against all <em>Pyrococcus</em> proteins with the MS-GF+ Search engine (v2016.10.26). The MS-GF+ parameter settings used are:</p>
<pre><code> -t 10ppm -ti 0,1 -tda 1 -m 3 -inst 1 -e 1 -protocol 0 -ntt 2 -minLength 6 -maxLength 30 
 -minCharge 2 -maxCharge 4 -n 1</code></pre>
<p>The modification file provide to MS-GF+ is:</p>
<pre><code>NumMods=2

57.021464,C,fix,any,Carbamidomethyl
15.994915,M,opt,any,Oxidation
H1O3P1,STY,opt,any,Phospho
</code></pre>
<p>For an explanation on how to use MS-GF+ and an explanation on all parameter settings, please read the MS-GF+ documentation at:</p>
<p><a href="https://omics.pnl.gov/software/ms-gf" class="uri">https://omics.pnl.gov/software/ms-gf</a></p>
<p>The MS-GF+ MZident output file from this search is included in this package for use in subsequent analysis steps.</p>
</div>
<div id="reading-a-ms-gf-mzident-output-file-" class="section level2">
<h2 class="hasAnchor">
<a href="#reading-a-ms-gf-mzident-output-file-" class="anchor"></a>Reading a MS-GF+ MZident output file.</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Load the saas library
<span class="kw">library</span>(saas)
## Load some convenience functions
<span class="kw">library</span>(dplyr, <span class="dt">quietly =</span> <span class="ot">TRUE</span>, <span class="dt">warn.conflicts =</span> <span class="ot">FALSE</span>) 

## Location of the zipped data files
zip_file_path =<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"extdata.zip"</span>, <span class="dt">package =</span> <span class="st">"saas"</span>)
## Unzip and get the (temporary) location of the mzid file with the MS-GF+ search results.
mzid_file_path =<span class="st"> </span><span class="kw">unzip</span>(zip_file_path, <span class="st">'pyrococcus.mzid'</span>,<span class="dt">exdir =</span> <span class="kw">tempdir</span>())

## Read and parse the mzid file
data =<span class="st"> </span><span class="kw"><a href="../reference/parse_msgf_mzid.html">parse_msgf_mzid</a></span>(mzid_file_path)

<span class="kw">glimpse</span>(data)</code></pre></div>
<pre><code>## Observations: 15,639
## Variables: 7
## $ spec_id       &lt;dbl&gt; 9834, 10918, 12207, 12179, 11387, 11027, 7833, 9...
## $ sequence      &lt;chr&gt; "GLEVSGYNCYIYPAMALAYGTSAIGAHHK", "MLVDSLGDIVITND...
## $ protein_id    &lt;chr&gt; "Q8U1K3|Formaldehyde:ferredoxin", "Q8TZL6|Thermo...
## $ score         &lt;dbl&gt; 2.383677e-34, 6.238274e-32, 2.329653e-31, 1.0498...
## $ database      &lt;chr&gt; "complete", "complete", "complete", "complete", ...
## $ decoy         &lt;lgl&gt; FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE,...
## $ database_size &lt;dbl&gt; 2051, 2051, 2051, 2051, 2051, 2051, 2051, 2051, ...</code></pre>
</div>
<div id="preprocessing-the-ms-gf-search-results-" class="section level2">
<h2 class="hasAnchor">
<a href="#preprocessing-the-ms-gf-search-results-" class="anchor"></a>Preprocessing the MS-GF+ search results.</h2>
<p>By default, the ‘preprocess()’ function removes all PSMs that assigned to both a decoy and target sequence. All rows in the dataframe that belong to the same PSMs (eg. one PSMs that match multiple proteins) are collapsed into one row.</p>
<p>When a path to a fasta file with the protein_ids from a subset of proteins in the fasta headers is provided, a new column is added that indicates if this PSM belongs to this subset of interest. These protein_ids should match the protein_ids from orginal fasta file used in the MS-GF+ search (and as indicated in the protein_id column in the data frame). This information is needed to use the all-sub method.</p>
<p>Optionally, it’s also possible to remove PSM’s that can be assigned to multiple proteins.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Unzip and get the (temporary) location of the file with fasta headers containing 
## the protein_ids from the protein subset of interest.
fasta_file_path =<span class="st"> </span><span class="kw">unzip</span>(zip_file_path, <span class="st">'transferase_activity_[GO:0016740].fasta'</span>, <span class="dt">exdir =</span> <span class="kw">tempdir</span>())

## Preprocess the data.
data_prep =<span class="st"> </span><span class="kw"><a href="../reference/preprocess.html">preprocess</a></span>(data, <span class="dt">is_subset =</span> fasta_file_path)

<span class="kw">glimpse</span>(data_prep)</code></pre></div>
<pre><code>## Observations: 15,107
## Variables: 9
## $ spec_id       &lt;dbl&gt; 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...
## $ sequence      &lt;chr&gt; "YEQIPWTQECSK", "NVQQKWFGK", "NLLESPKIK", "KHLEQ...
## $ decoy         &lt;lgl&gt; FALSE, TRUE, TRUE, FALSE, FALSE, FALSE, TRUE, FA...
## $ database      &lt;chr&gt; "complete", "complete", "complete", "complete", ...
## $ database_size &lt;dbl&gt; 2051, 2051, 2051, 2051, 2051, 2051, 2051, 2051, ...
## $ score         &lt;dbl&gt; 2.560707e-05, 1.458271e-04, 9.241630e-06, 1.7355...
## $ subset        &lt;lgl&gt; FALSE, TRUE, FALSE, FALSE, FALSE, FALSE, FALSE, ...
## $ non_subset    &lt;lgl&gt; TRUE, FALSE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE,...
## $ protein_id    &lt;chr&gt; "Q8U285|Uncharacterized", "Q8U168|Glycosyl", "P8...</code></pre>
</div>
<div id="evaluating-the-decoy-psm-distribution-" class="section level2">
<h2 class="hasAnchor">
<a href="#evaluating-the-decoy-psm-distribution-" class="anchor"></a>Evaluating the decoy PSM distribution.</h2>
<p>The implementation of the all-sub in this packages adopt the target-decoy approach for FDR estimation. A subset FDR can be obtained by applying TDA on the subset target and decoy PSMs.</p>
<p>However, in very small subsets with little subset target and decoy PSMs, there can be a large sample-to-sample variability on the FDR estimates. We provide extensions to the basic subset TDA FDR to obtain more stable FDR estimates. Key is the use of a large set of decoy PSMs to reliable estimate the distribution of incorrect target PSM scores. By default we use all decoys from the complete search (subset and non-subset PSMs). The user can also define another set of extra decoys and use this for FDR calculation (see <a href="#example2">example 2</a>).</p>
<p>Bot the classical TDA FDR and our more stable TDA FDR relie on the assumption that the distribution of incorrect target PSM scores can be approximated by the decoy PSM score distribution.</p>
<p>In this package we provide diagnostic plots to verify the assumption that the decoy PSM score distribution follows the incorrect target PSM distribution. Tthe ‘score_higher’ parameter in ‘plot_diag()’ indicates if higher score values mean a more confident PSM. We take the MS-GF+ SpecEValue as the PSM score for FDR calculation and smaller SpecEValue indicate a better match. Therefore we have to set the ‘score_higher’ parameter to ‘FALSE’. Note that in the manuscript we used <span class="math inline">\(- log (\)</span>MS-GF+ SpecEValue<span class="math inline">\()\)</span> as the score for visualisation purposes. Please consult the documentation of your search engine of choice on how to interpret the PSM score.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">diagnostics =<span class="st"> </span><span class="kw"><a href="../reference/plot_diag.html">plot_diag</a></span>(data_prep, <span class="dt">score_higher =</span> <span class="ot">FALSE</span>)
diagnostics<span class="op">$</span>all</code></pre></div>
<p><img src="saas_files/figure-html/unnamed-chunk-3-1.png" width="700"><strong>Panel a</strong> shows the posterior distribution of pi_0 given the observed number of target and decoy PSMs in the subset. The vertical line indicates the conservative pi_0 estimate used in the calculations. At very high <span class="math inline">\(\pi_0\)</span> uncertainty (broad peak), you can also opt to use the Benjamine Hochberg procedure to minimize sample to sample variability which would mean that pi_0 would be set to 1 (see <strong>FDR_BH</strong> in the next section). However, this will come at the expense of too conservative PSM lists.</p>
<p>The distributional assumption for the decoys can be verified through a PP-plot where the empirical Cumulative Distribution Function (eCDF) of the decoys is plotted against the eCDF of the subset target PSMs. The PP-plots in <strong>panel b - d</strong> display the target subset PSMs plotted against all decoy PSMs from the complete search, the decoy subset PSMs plotted against all decoy PSMs from the complete search, and the target subset PSMs plotted against the decoy PSMs from the complete search, respectively. The full line in panel <strong>b</strong> and <strong>d</strong> indicates a line with a slope of <span class="math inline">\(\pi_0\)</span>. The full line in panel <strong>c</strong> indicates the identity line. When the distributional assumption holds then the first part of the plot in <strong>b</strong> and <strong>d</strong> should be linear with a slope that equals <span class="math inline">\(\pi_0\)</span>. The second part of the plot deviates from the line towards higher percentiles and will ultimately become vertical (decoy percentile = 1). If we see this profile in panel <strong>b</strong>, we have a good indication that the set of decoys from the complete search is representative for the mixture component for incorrect PSMs of the target mixture distribution. Deviations from this pattern might be subtle, therefore we provide the PP plots in <strong>c</strong> and <strong>d</strong> to support the conclusion drawn from panel <strong>b</strong>. When there is high uncertainty on pi_0 as indicated by <strong>a</strong>, then the linear pattern in the data points might deviate from the drawn solid line, but should still be more or less linear. The PP-plot in panel <strong>c</strong> shows the subset decoy PSMs plotted against all decoy PSMs. The whole plot should follow the identity line, indicating that the complete set of decoys is a good representation of the subset decoys. To verify that the subset decoys are representative for the mixture component for incorrect PSMs of the target mixture distribution, we look at the PP-plot of the subset decoys against the subset targets in panel <strong>d</strong>. The profile should look as described for panel <strong>b</strong>. If the profile matches in panel <strong>d</strong> but does not for panel <strong>b</strong>, then we suggest to not use the extra decoy set and use only the subset decoys for FDR estimation. When the profile does not match in panel <strong>d</strong>, the subset decoys might not be representative for incorrect PSMs. This can indicate that <span class="math inline">\(\pi_0\)</span> is estimated incorrectly, since this is based on the subset PSM scores. In this case, the first part of the plot in panel <strong>d</strong> can deviate from the (incorrect) <span class="math inline">\(\pi_0\)</span> slope line. But if this first part is linear, it still indicates that the extra set of decoys is representative for the mixture component of incorrect target PSMs. Since <span class="math inline">\(\pi_0\)</span> is unknow in this case we set <span class="math inline">\(\pi_0\)</span> to 1 (see <strong>FDR_BH</strong> in the next section).</p>
<p>We can conclude from the above diagnostic plots that the large decoy set is an appropiate candiate set to estimate the incorrect subset target PSM distribution.</p>
</div>
<div id="estimating-fdr-on-the-subset-of-interest-" class="section level2">
<h2 class="hasAnchor">
<a href="#estimating-fdr-on-the-subset-of-interest-" class="anchor"></a>Estimating FDR on the subset of interest.</h2>
<p>The output from ‘preprocess()’ can be immediatly used to estimate the FDR in the PSM subset of interest.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">data_results =<span class="st"> </span><span class="kw"><a href="../reference/calculate_fdr.html">calculate_fdr</a></span>(data_prep, <span class="dt">score_higher =</span> <span class="ot">FALSE</span>)
<span class="kw">glimpse</span>(data_results)</code></pre></div>
<pre><code>## Observations: 15,107
## Variables: 13
## $ spec_id       &lt;dbl&gt; 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...
## $ sequence      &lt;chr&gt; "YEQIPWTQECSK", "NVQQKWFGK", "NLLESPKIK", "KHLEQ...
## $ decoy         &lt;lgl&gt; FALSE, TRUE, TRUE, FALSE, FALSE, FALSE, TRUE, FA...
## $ database      &lt;chr&gt; "complete", "complete", "complete", "complete", ...
## $ database_size &lt;dbl&gt; 2051, 2051, 2051, 2051, 2051, 2051, 2051, 2051, ...
## $ score         &lt;dbl&gt; 2.560707e-05, 1.458271e-04, 9.241630e-06, 1.7355...
## $ subset        &lt;lgl&gt; FALSE, TRUE, FALSE, FALSE, FALSE, FALSE, FALSE, ...
## $ non_subset    &lt;lgl&gt; TRUE, FALSE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE,...
## $ protein_id    &lt;chr&gt; "Q8U285|Uncharacterized", "Q8U168|Glycosyl", "P8...
## $ pi_0_cons     &lt;dbl&gt; 0.3653846, 0.3653846, 0.3653846, 0.3653846, 0.36...
## $ FDR           &lt;dbl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, ...
## $ FDR_BH        &lt;dbl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, ...
## $ FDR_stable    &lt;dbl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, ...</code></pre>
<p>We can choose in the output from three different all-sub FDR estimations: The estimated stable FDR at this score cutoff for subset PSMs.</p>
<ol style="list-style-type: decimal">
<li>FDR: Calculated according the classical TDA method. Does not work well small subsets.</li>
<li>FDR_stable: Our improved FDR method which is more stable in small subsets. For large subsets, <em>FDR_stable</em> estimates will be close to <strong>FDR</strong> estimates.</li>
<li>FDR_BH: Bejamini Hochberg FDR procedure. Use this when you have a large decoy set but no decoy information on the subset PSMs (e.g. when the search engine does not return decoy protein ids). This FDR estimate is more conservative then <strong>FDR</strong> and <strong>FDR_stable</strong>.</li>
</ol>
<p>The ‘transferase activity’ subset is rather small with 104 target and 37 decoy PSMs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">count</span>(data_results, subset, decoy)</code></pre></div>
<pre><code>## Source: local data frame [4 x 3]
## Groups: subset [?]
## 
##   subset decoy     n
##    &lt;lgl&gt; &lt;lgl&gt; &lt;int&gt;
## 1  FALSE FALSE 13158
## 2  FALSE  TRUE  1808
## 3   TRUE FALSE   104
## 4   TRUE  TRUE    37</code></pre>
<p>From the diagnostic plots we concluded that the set of decoy PSMs from the complete search is a good candidate set to approximate the incorrect target PSM distribution.</p>
<p>So, in this case it’s recommended to use the more stable FDR estimates in <strong>FDR_stable</strong>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">results_1_FDR =<span class="st"> </span><span class="kw">filter</span>(data_results, subset, <span class="op">!</span>decoy, FDR_stable <span class="op">&gt;=</span><span class="st"> </span>.<span class="dv">01</span>) 
<span class="kw">glimpse</span>(results_1_FDR)</code></pre></div>
<pre><code>## Observations: 31
## Variables: 13
## $ spec_id       &lt;dbl&gt; 85, 1361, 3519, 6026, 6857, 7516, 8462, 8490, 95...
## $ sequence      &lt;chr&gt; "YKGRGFRK", "KLGENNER", "FGTFLLTILTK", "YKFKNNYR...
## $ decoy         &lt;lgl&gt; FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE,...
## $ database      &lt;chr&gt; "complete", "complete", "complete", "complete", ...
## $ database_size &lt;dbl&gt; 2051, 2051, 2051, 2051, 2051, 2051, 2051, 2051, ...
## $ score         &lt;dbl&gt; 5.231224e-05, 3.140562e-05, 7.486767e-06, 4.2052...
## $ subset        &lt;lgl&gt; TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, ...
## $ non_subset    &lt;lgl&gt; FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE,...
## $ protein_id    &lt;chr&gt; "Q8U164|Glycosyl", "Q8U4L8|6,7-dimethyl-8-ribity...
## $ pi_0_cons     &lt;dbl&gt; 0.3653846, 0.3653846, 0.3653846, 0.3653846, 0.36...
## $ FDR           &lt;dbl&gt; 0.23333333, 0.17441860, 0.07317073, 0.21348315, ...
## $ FDR_BH        &lt;dbl&gt; 0.63759109, 0.52501418, 0.21983740, 0.59408666, ...
## $ FDR_stable    &lt;dbl&gt; 0.23296597, 0.19183210, 0.08032520, 0.21707013, ...</code></pre>
<p>At 1% FDR, we return 31 PSMs.</p>
</div>
</div>
<div id="example2" class="section level1">
<h1 class="hasAnchor">
<a href="#example2" class="anchor"></a>Example 2: Provide an external set of decoys</h1>
</div>
<div id="sessioninfo" class="section level1">
<h1 class="hasAnchor">
<a href="#sessioninfo" class="anchor"></a>sessionInfo()</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sessionInfo</span>()</code></pre></div>
<pre><code>## R version 3.3.3 (2017-03-06)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Arch Linux
## 
## locale:
## [1] C
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] dplyr_0.5.0     saas_0.0.0.9000
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_0.12.10        plyr_1.8.4          ProtGenerics_1.6.0 
##  [4] tools_3.3.3         digest_0.6.12       evaluate_0.10      
##  [7] tibble_1.3.0        gtable_0.2.0        shiny_1.0.0        
## [10] DBI_0.6             yaml_2.1.14         parallel_3.3.3     
## [13] stringr_1.2.0       knitr_1.15.1        hms_0.3            
## [16] rprojroot_1.2       grid_3.3.3          cowplot_0.7.0      
## [19] Biobase_2.34.0      R6_2.2.0            rmarkdown_1.4      
## [22] ggplot2_2.2.1       readr_1.1.0         mzR_2.8.0          
## [25] magrittr_1.5        backports_1.0.5     scales_0.4.1       
## [28] codetools_0.2-15    htmltools_0.3.5     BiocGenerics_0.20.0
## [31] assertthat_0.2.0    mime_0.5            colorspace_1.3-2   
## [34] xtable_1.8-2        httpuv_1.3.3        labeling_0.3       
## [37] stringi_1.1.5       lazyeval_0.2.0.9000 munsell_0.4.3      
## [40] markdown_0.7.7</code></pre>
<div id="refs" class="references">
<div id="ref-BenjaminiHochberg1995">
<p>Benjamini, Yoav, and Yosef Hochberg. 1995. “Controlling the False Discovery Rate: A Practical and Powerful Approach to Multiple Testing.” <em>Journal of the Royal Statistical Society: Series B</em> 57 (1). Blackwell Publishing for the Royal Statistical Society: 289–300.</p>
</div>
<div id="ref-Efron2008a">
<p>Efron, Bradley. 2008. “Microarrays, Empirical Bayes and the Two-Groups Model.” <em>Statistical Science</em> 23 (1): 1–22. doi:<a href="https://doi.org/10.1214/07-STS236">10.1214/07-STS236</a>.</p>
</div>
<div id="ref-Elias2007">
<p>Elias, Joshua E, and Steven P Gygi. 2007. “Target-decoy search strategy for increased confidence in large-scale protein identifications by mass spectrometry.” <em>Nature Methods</em> 4 (3): 207–14. doi:<a href="https://doi.org/10.1038/nmeth1019">10.1038/nmeth1019</a>.</p>
</div>
<div id="ref-Sticker2017">
<p>Sticker, Adriaan, Lieven Clement, and Lennart Martens. 2017. “Mass Spectrometrists Should Search for All Peptides, but Assess Only the Ones They Care About.” <em>bioRxiv</em>. Cold Spring Harbor Labs Journals. doi:<a href="https://doi.org/10.1101/094581">10.1101/094581</a>.</p>
</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#introduction">Introduction</a><ul class="nav nav-pills nav-stacked">
<li><a href="#false-discovery-rate-fdr-estimation">False discovery rate (FDR) estimation</a></li>
      <li><a href="#target-decoy-approach">Target-decoy approach</a></li>
      <li><a href="#stable-target-decoy-approach-for-subsets">Stable target-decoy approach for subsets</a></li>
      </ul>
</li>
      <li>
<a href="#example-data-analysis">Example data analysis</a><ul class="nav nav-pills nav-stacked">
<li><a href="#the-data">The data</a></li>
      <li><a href="#database-search">Database search</a></li>
      <li><a href="#reading-a-ms-gf-mzident-output-file.">Reading a MS-GF+ MZident output file.</a></li>
      <li><a href="#preprocessing-the-ms-gf-search-results.">Preprocessing the MS-GF+ search results.</a></li>
      <li><a href="#evaluating-the-decoy-psm-distribution.">Evaluating the decoy PSM distribution.</a></li>
      <li><a href="#estimating-fdr-on-the-subset-of-interest.">Estimating FDR on the subset of interest.</a></li>
      </ul>
</li>
      <li><a href="#example2">Example 2: Provide an external set of decoys</a></li>
      <li><a href="#sessioninfo">sessionInfo()</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Adriaan Sticker.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
